File: HW3JS/Car/vehicle.js
"use strict"

class BadVehicleState {
    constructor(issue) {
        this.issue = issue;
    }
}
class VehicularVehicle{
    constructor(engine, owner, color) {
        console.log("Creating Vehicle");
        this._engine = engine;
        this._owner = owner;
        this._color = color;
    }
    operate(){
        throw new Error("operating abstract vehicle...");
    }
    get owner(){
        return this._owner;
    }
    get color(){
        return this._color;
    }
    get engine(){
         return this._engine;
    }
 }


class Car extends VehicularVehicle{
    constructor(engine, owner, color) {
        super(engine, owner, color);
        console.log("Signing the papers for your car..");
        this.haskeys = true;
    }

    drive(){
        if (this._haskeys == true)
            console.log(this.owner + " is driving a " + this.color + " " + this.engine);
        else {
            throw Car.BAD_CAR_OWNER_DRIVE;
        }
    }
    get owner() {
        return this._owner;
    }
    get color(){
        return this._color;
    }
    get engine() {
        return this._engine;
    }
    set owner (own)
    {
        if(typeof own == "string"){
            this.owner = own;
        }
        else{
            throw Car.BAD_CAR_OWNER_NOTSET;
        }
    }
    get haskeys(){
        return this._haskeys;
    }
    set haskeys(key){
        if(typeof key == "boolean"){
            this._haskeys = key;
        }
        else
            throw Car.BAD_CAR_KEYS_NOTMATCHED;
    }

}

Car.BAD_CAR_OWNER_DRIVE = new BadVehicleState("Driving a car that isnt yours is WRONG!");
Car.BAD_CAR_OWNER_NOTSET = new BadVehicleState("Someone must own the car!?!");
Car.BAD_CAR_KEYS_NOTMATCHED = new BadVehicleState("Keys must be set!");

exports.Car = Car;
exports.BadVehicleState = BadVehicleState;

File: HW6/src/Vehicle.cpp
#include "../include/Vehicle.h"

#include <iostream>

namespace vehicle{

  Vehicle::Vehicle(Color _color, Owner _owner)
  : color(_color), owner(_owner), keys(true){
    std::cout << "Signing papers and creating Abstract Vehicle, giving " << _owner << " the keys" << std::endl;
  }
  Vehicle::~Vehicle(){
    std::cout << "Selling vehicle and salvaging parts..." << std::endl;
  }
  void Vehicle::setOwner(Owner _owner){
    owner = _owner;
  }
  Owner Vehicle::getOwner() const{
    return owner;
  }
  Color Vehicle::getColor() const{
    return color;
  }
  void Vehicle::setColor(Color _color){
    color = _color;
  }
  bool Vehicle::setKeys(bool _keys){
    keys = _keys;
  }
  bool Vehicle::getKeys() const{
    return keys;
  }
  void Vehicle::drive(){
    if(getKeys() == true){
    std::cout << getOwner() << " is driving Abstract " << getColor() << " Vehicle" << std::endl;
    }
  }

}
